apps/hash-ai-worker-ts/src/activities/shared/get-llm-response.ts

```diff
@@ -666,15 +689,170 @@ const getOpenAiResponse = async (
  * `model` provided in the parameters.
  */
 export const getLlmResponse = async <T extends LlmParams>(
-  params: T,
+  params: T & {
+    /**
+     * Required for tracking usage on a per-user basis.
+     *
+     * @todo: consider abstracting this in a wrapper method, or via
+     * generic params (via a `logUsage` method).
+     */
+    userAccountId: AccountId;
+    graphApiClient: GraphApi;
+    linkUsageRecordToEntities: {
+      linkEntityTypeId: [
+        typeof systemLinkEntityTypes.incurredIn.linkEntityTypeId,
+        typeof systemLinkEntityTypes.created.linkEntityTypeId,
+        typeof systemLinkEntityTypes.updated.linkEntityTypeId,
+      ][number];
+      entityId: EntityId;
+    }[];
+  },
 ): Promise<LlmResponse<T>> => {
-  if (isLlmParamsAnthropicLlmParams(params)) {
-    const response = await getAnthropicResponse(params);
+  const {
+    userAccountId,
+    graphApiClient,
+    linkUsageRecordToEntities,
+    ...remainingLlmParams
+  } = params;
 
-    return response as unknown as LlmResponse<T>;
-  } else {
-    const response = await getOpenAiResponse(params);
+  /**
+   * Check whether the user has exceeded their usage limit, before
+   * proceeding with the LLM request.
+   */
+  const userHasExceededUsageStatus =
+    await userExceededServiceUsageLimitActivity({
+      graphApiClient,
+      userAccountId,
+    });
+
+  if (userHasExceededUsageStatus.code !== StatusCode.Ok) {
+    return {
+      status: "exceeded-usage-limit",
+      message:
+        userHasExceededUsageStatus.message ??
+        "You have exceeded your usage limit.",
+    };
+  }
 
-    return response as unknown as LlmResponse<T>;
+  let aiAssistantAccountId: AccountId;
+  try {
+    aiAssistantAccountId = await getMachineActorId(
+      { graphApi: graphApiClient },
+      { actorId: userAccountId },
+      { identifier: "hash-ai" },
+    );
+  } catch {
+    return {
+      status: "internal-error",
+      message: `Failed to retrieve AI assistant account ID ${userAccountId}`,
+    };
```

Comment:
This doesn't use `getAiAssistantAccountIdActivity`, which is used in `infer-entities-from-content-action` and `infer-entities` and ensures that the AI Assistant has permission on the specified web. 

It's possible that this code will be executed without having gone through either of those files, which means the AI Assistant may not have been granted permission on the web and creating the usage records will fail. 

It needn't be checked in `infer-entities-from-content-action` since there's nothing in there affecting the web, though it should be here.

We could also consider getting rid of the AI Assistant bot altogether and just use the appropriate web bot, which will have permission. The AI Assistant bot was mainly to be able to identify which things came from AI, but we are covering that in the new provenance system. This would save yet another check in this function, which we'll be calling quite a lot and now has a lot of hoops to jump through (check usage, get AI bot, check AI bot permissions)

This is also linked to how we attribute usage (e.g. are we creating usage records in an org web in some circumstances).

Reply:
Thanks for catching this, I've made use of the method in https://github.com/hashintel/hash/pull/4379/commits/7641eddd58f8d7d9b5b57203a2d119d469d7d71a. I've set it to grant permissions in the user's web for now, but this might require changing if we allow setting the `webId` that the usage record is created in.

Also added a todo for using the machine actor instead once we have this in the provenance metadata.